---
title: "HJ Andrews Synoptic: Bacterial Analysis"
author: "Nathan I. Wisnoski"
date: "11/16/2019"
output: html_document
---

# Initial setup 
```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE, 
  fig.align = "center", 
  fig.width = 6,
  message = FALSE,
  warning = FALSE
)

library("vegan")
library("tidyverse")
library("cowplot")
library("ggrepel")
library("readxl")
library("ggridges")
library("ftmsRanalysis")
library("viridis")

theme_set(theme_bw() +
            theme(axis.title = element_text(size = 12),
                  axis.text = element_text(size = 10),
                  legend.position = "top", 
                  strip.text = element_text(size = 12),
                  strip.background = element_blank(),
                  legend.text = element_text(size = 10),
                  legend.title = element_text(size = 12),
                  panel.grid = element_blank()))


# load external functions
source("analysis/mothur_tools.R")

```

# Load environmental data
First, we load the data. This includes the site-by-species matrix (generated in Mothur, v. 1.41.1), the RDP taxonomy, the environmental data, and the phylogenetic tree (generated with FastTreeMP).
```{r load, echo=FALSE, message=FALSE}
## Import Shared, Design, and Environment Files

# Define Inputs
# Design = general design file for experiment
# shared = OTU table from mothur with sequence similarity clustering
# Taxonomy = Taxonomic information for each OTU

# Import Design
#design.total <- read.delim("data/design.txt", header=T, row.names=1)

# Import Shared Files
OTUs <- read.otu(shared = "data/hja-synoptic.0.03.shared", cutoff = "0.03") # 97% Similarity
OTUs <- OTUs[rownames(OTUs) != "hja2016_046",] # remove these two sites
OTUs <- OTUs[rownames(OTUs) != "hja2016_115",]

# Import Taxonomy
OTU.tax <- read.tax(taxonomy = "data/hja-synoptic.0.03.taxonomy", format = "rdp")

# Import Env
options(readr.default_locale=readr::locale(tz="America/Los_Angeles"))

# Some NAs here due to non-recorded time points -- ok
env.aquatic <- read_csv("data/hja-env_data_clean.csv",
                      col_types = cols(
                        Date = col_date(format = "%Y-%m-%d"),
                        Time = col_time(format = "%H:%M")))
#env.aquatic$Site.Code <- env.aquatic$`Site Code`

env.soils <- read_csv("data/hja-synoptic_env-data-soils.csv",
                      col_types = cols(
                        date = col_date()))

design <- read.csv("data/hja-synoptic_sequence-sample-list.csv")
rownames(design) <- design$Sample.Name
```

# Load microbial data
```{r}
dim(OTUs)
sort(rowSums(OTUs))
hist(log10(rowSums(OTUs)), breaks = 30)

# Filter samples with fewer than 5000 reads (removes 10 sites)
OTUs <- OTUs[which(rowSums(OTUs)>5000),]

# check if order is the same
design <- subset(design, Sample.Name %in% rownames(OTUs))
sum(rownames(OTUs) != rownames(design))

set.seed(47405)
OTUs <- rrarefy(OTUs, min(rowSums(OTUs)))

OTUs <- OTUs[,which(colSums(OTUs) != 0)]
OTUs.hel <- decostand(OTUs, method = "hel")
```

# Load FTICRMS data
```{r}
fticrms <- read_xlsx("data/hja-FTICRMS.xlsx")

# peak_data_long <- fticrms[,c(1,15:76)] %>% 
#   pivot_longer(cols = !(`Measured Ionic Mass (m/z)`),
#                names_to = "Site.Code", 
#                values_to = "Count") %>% 
#   rename(mass_ratio = `Measured Ionic Mass (m/z)`)

peak_data <- fticrms[,c(1,15:76)] %>% 
  rename(mass_ratio = `Measured Ionic Mass (m/z)`) %>% 
  as.data.frame(.)
sites <- data.frame(Site.Code = as.character(colnames(peak_data)[-1]))
sites$sample_type <- c(rep("sediment", nrow(sites)-2), rep("labqc", 2))

# sites <- env.aquatic %>% filter(`Sample Type` == "Sediment") %>% 
#   rename(Site.Code = `Site Code`) %>% 
#   right_join(sites, by = "Site.Code") %>% 
#   .[,c(1:4)]

meta_df <-fticrms[,1:13] %>% 
  rename(mass_ratio = `Measured Ionic Mass (m/z)`) %>% 
  as.data.frame(.)

hja_peakdat <- as.peakData(e_data = peak_data, 
            f_data = sites,
            e_meta = meta_df, 
            edata_cname = "mass_ratio",
            fdata_cname = "Site.Code",
            mass_cname = "mass_ratio",
            instrument_type = "12T", 
            isotopic_cname = "C13",
            isotopic_notation = "1",
            c_cname = "C",
            h_cname = "H",
            o_cname = "O",
            n_cname = "N",
            s_cname = "S",
            p_cname = "P")

summary(hja_peakdat)

filter_obj <- ftmsRanalysis::mass_filter(hja_peakdat)
hja_peakdat_filtered <- applyFilt(filter_obj, hja_peakdat, min_mass = 200, max_mass = 900)
isotope_filter <- ftmsRanalysis::emeta_filter(hja_peakdat_filtered, cname = "C13")
hja_peakdat_filtered <- applyFilt(isotope_filter, hja_peakdat_filtered, max_val = 0 )

hja_peakdat_pa <- edata_transform(hja_peakdat_filtered, data_scale = "pres")
plot(hja_peakdat_pa)
hja_peakdat_pa <- compound_calcs(hja_peakdat_pa)
hja_peakdat_pa <- assign_elemental_composition(hja_peakdat_pa)
table(hja_peakdat_pa$e_meta[,getElCompColName(hja_peakdat_pa)])
hja_peakdat_pa <- assign_class(hja_peakdat_pa)
table(hja_peakdat_pa$e_meta[,getBS1ColName(hja_peakdat_pa)])

hja_peakdat_pa <- applyFilt(molecule_filter(hja_peakdat_pa), hja_peakdat_pa, min_number=2)
hja_peakdat_pa <- applyFilt(formula_filter(hja_peakdat_pa), hja_peakdat_pa)

hja_peakdat_pa <- subset(hja_peakdat_pa, samples = sites$Site.Code[which(sites$sample_type!="labqc")], check_rows = T)
plot(hja_peakdat_pa)
summary(hja_peakdat_pa)

one_sample <- subset(hja_peakdat_pa, samples = "CC-1")
vanKrevelenPlot(one_sample)
densityPlot(hja_peakdat_pa, variable = "GFE", curve_colors = viridis(62))
```

# Analysis

## PCoA
```{r}
# Visualize 
OTUs.hel.dist <- vegdist(OTUs.hel, method = "euclid")
hel.pcoa <- cmdscale(OTUs.hel.dist, k = 3, eig = T)
hja.evals <- eigenvals(hel.pcoa)
var1 <- hja.evals[1]/sum(hja.evals) * 100
var2 <- hja.evals[2]/sum(hja.evals) * 100
var3 <- hja.evals[3]/sum(hja.evals) * 100
sum(var1, var2, var3)

pcdims <- hel.pcoa$points[,1:3]
colnames(pcdims) <- c("PCoA1", "PCoA2", "PCoA3")

# Plot 
left_join(rownames_to_column(as.data.frame(pcdims), var = "Sample.Name"), design) %>% 
  ggplot(aes(x = PCoA1, y = PCoA3, color = Sample.Type)) + 
  geom_point(alpha = 0.5) + 
  stat_ellipse() +
  coord_fixed() +
  #geom_text_repel(aes(label = Site.Code)) +
  labs(x = paste0("PCoA1 (", round(var1,2),"%)"),
       y = paste0("PCoA3 (", round(var3,2),"%)")) +
  scale_color_brewer("", palette = "Dark2") +
  ggsave("figures/hja-synoptic-ordination.png", dpi = 600, width = 6, height = 6, units = "in")
```

## PCoA molecules
```{r}
hja_peaks <- hja_peakdat_pa$e_data
rownames(hja_peaks) <- hja_peaks$mass_ratio
hja_peaks <- t(hja_peaks[,-1])
head(hja_peaks[,1:10])
hja_peaks_dist <- vegdist(hja_peaks, method = "bray")

hja_peaks_pcoa <- cmdscale(hja_peaks_dist, k = 3, eig = T)
hja_peaks_evals <- eigenvals(hja_peaks_pcoa)
var1 <- hja_peaks_evals[1]/sum(hja_peaks_evals) * 100
var2 <- hja_peaks_evals[2]/sum(hja_peaks_evals) * 100
var3 <- hja_peaks_evals[3]/sum(hja_peaks_evals) * 100
sum(var1, var2, var3)

pcdims <- hja_peaks_pcoa$points[,1:3]
colnames(pcdims) <- c("PCoA1", "PCoA2", "PCoA3")

# Plot 
left_join(rownames_to_column(as.data.frame(pcdims), var = "Site.Code"), 
          env.aquatic %>% filter(`Sample Type` == "Sediment") %>% rename(Site.Code = `Site Code`),
          by = "Site.Code") %>% 
  ggplot(aes(x = PCoA1, y = PCoA2, color = log10(`Drainage_area_ha`))) + 
  geom_point(alpha = 0.5) + 
  #stat_ellipse() +
  coord_fixed() +
  geom_text_repel(aes(label = Site.Code)) +
  labs(x = paste0("PCoA1 (", round(var1,2),"%)"),
       y = paste0("PCoA2 (", round(var2,2),"%)")) 

```

```{r}
sed_sitecodes_in_seqs <- subset(design, Sample.Type == "Sediment")$Site.Code


# do sed fticrms pcoa
sed_peaks_pa <- subset(hja_peakdat_pa, samples = sed_sitecodes_in_seqs)
sed_peaks <- sed_peaks_pa$e_data
rownames(sed_peaks) <- sed_peaks$mass_ratio
sed_peaks <- t(sed_peaks[,-1])
head(sed_peaks[,1:10])
sed_peaks_dist <- vegdist(sed_peaks, method = "bray")

sed_peaks_pcoa <- cmdscale(sed_peaks_dist, k = 3, eig = T)
sed_peaks_evals <- eigenvals(sed_peaks_pcoa)
var1 <- sed_peaks_evals[1]/sum(sed_peaks_evals) * 100
var2 <- sed_peaks_evals[2]/sum(sed_peaks_evals) * 100
var3 <- sed_peaks_evals[3]/sum(sed_peaks_evals) * 100
sum(var1, var2, var3)

pcdims <- sed_peaks_pcoa$points[,1:3]
colnames(pcdims) <- c("PCoA1", "PCoA2", "PCoA3")

# Plot 
left_join(rownames_to_column(as.data.frame(pcdims), var = "Site.Code"), 
          env.aquatic %>% filter(`Sample Type` == "Sediment") %>% rename(Site.Code = `Site Code`),
          by = "Site.Code") %>% 
  ggplot(aes(x = PCoA1, y = PCoA2, color = log10(`Drainage_area_ha`))) + 
  geom_point(alpha = 0.5) + 
  #stat_ellipse() +
  coord_fixed() +
  geom_text_repel(aes(label = Site.Code)) +
  labs(x = paste0("PCoA1 (", round(var1,2),"%)"),
       y = paste0("PCoA2 (", round(var2,2),"%)")) +
  ggsave("figures/sed_molecular_pcoa.png", dpi = 600, width = 6, height = 3/4*6)




# do sed microbe pcoa
sed_samples_in_ICRMS <- subset(design, Sample.Type == "Sediment" & Site.Code %in% sed_peaks_pa$f_data$Site.Code)$Sample.Name

sed_otus_dist <- vegdist(OTUs[sed_samples_in_ICRMS,], method = "bray")
sed_pcoa <- cmdscale(sed_otus_dist, k = 3, eig = T)
sed_evals <- eigenvals(sed_pcoa)
var1 <- sed_evals[1]/sum(sed_evals) * 100
var2 <- sed_evals[2]/sum(sed_evals) * 100
var3 <- sed_evals[3]/sum(sed_evals) * 100
sum(var1, var2, var3)

pcdims <- sed_pcoa$points[,1:3]
colnames(pcdims) <- c("PCoA1", "PCoA2", "PCoA3")

left_join(rownames_to_column(as.data.frame(pcdims), var = "Sample.Name"), design) %>% 
  ggplot(aes(x = PCoA1, y = PCoA3)) + 
  geom_point(alpha = 0.5) + 
  coord_fixed() +
  geom_text_repel(aes(label = Site.Code), max.overlaps = 20) +
  labs(x = paste0("PCoA1 (", round(var1,2),"%)"),
       y = paste0("PCoA3 (", round(var3,2),"%)")) +
  scale_color_brewer("", palette = "Dark2") +
  ggsave("figures/sed_microbial_pcoa.png", dpi = 600, width = 6, height = 3/4*6)


scores(sed_pcoa)
scores(sed_peaks_pcoa)

sed_procrust <- protest(sed_pcoa, sed_peaks_pcoa)
plot(sed_procrust)
```










##
```{r}

env <- left_join(design, env.soils) %>% left_join(env.aquatic)

soil.dat <- inner_join(design, env.soils)
soil.coords <- SoDA::geoXY(soil.dat$latitude, soil.dat$longitude)
soil.geo <- dist(soil.coords)
soil.otus <- OTUs.hel[which(design$Sample.Type == "Terrestrial soil"),]
soil.dists <- vegdist(soil.otus, method = "euclid")
plot(soil.geo, soil.dists)
cbind(geo = soil.geo, com = soil.dists) %>% as_tibble() %>% 
  ggplot(aes(x = geo, y = com)) +
  geom_point() +
  geom_smooth(method = 'lm')

sed.dat <- inner_join(design[which(design$Sample.Type == "Sediment"),], (env.aquatic))
sed.geo <- dist(SoDA::geoXY(aquatic.dat$latitude, aquatic.dat$longitude))
sed.otus <- OTUs.hel[sed.dat$Sample.Name,]
sed.dists <- vegdist(sed.otus, method = "euclid")
cbind(geo = sed.geo, com = sed.dists) %>% as_tibble() %>% 
  ggplot(aes(x = geo, y = com)) +
  geom_point() +
  geom_smooth(method = 'lm')

aquatic.dat <- inner_join(design[which(design$Sample.Type == "Surface water"),], (env.aquatic))
aquatic.geo <- dist(SoDA::geoXY(aquatic.dat$latitude, aquatic.dat$longitude))
aquatic.otus <- OTUs.hel[aquatic.dat$Sample.Name,]
aquatic.dists <- vegdist(aquatic.otus, method = "euclid")
cbind(geo = aquatic.geo, com = aquatic.dists) %>% as_tibble() %>% 
  ggplot(aes(x = geo, y = com)) +
  geom_point() +
  geom_smooth(method = 'lm')


hypor.dat <- inner_join(design[which(design$Sample.Type == "Hyporheic water"),], (env.aquatic))
hypor.geo <- dist(SoDA::geoXY(hypor.dat$latitude, hypor.dat$longitude))
hypor.otus <- OTUs.hel[hypor.dat$Sample.Name,]
hypor.dists <- vegdist(hypor.otus, method = "euclid")
cbind(geo = hypor.geo, com = hypor.dists) %>% as_tibble() %>% 
  ggplot(aes(x = geo, y = com)) +
  geom_point() +
  geom_smooth(method = 'lm')

all.dat <- rbind.data.frame(
  inner_join(design, select(env.aquatic, Site.Code, latitude, longitude)),
  inner_join(design, select(env.soils, Site.Code, latitude, longitude)))
all.geo <- dist(SoDA::geoXY(all.dat$latitude, all.dat$longitude))
all.otus <- OTUs.hel[droplevels(all.dat$Sample.Name),]
all.dists <- vegdist(all.otus, method = "euclid")
all.dists.df <- simba::liste(all.dists, entry = "com")
all.dists.df <- all.dists.df %>% 
  mutate(site1 = as.character(NBX), site2 = as.character(NBY),
         hab1 = design[site1,"Sample.Type"],
         hab2 = design[site2,"Sample.Type"],
         habs = paste(hab1, hab2, sep = "-")) %>% 
  mutate(geo = all.geo)
for(i in 1:nrow(all.dists.df)){
  hab1 = as.character(all.dists.df$hab1[i])
  hab2 = as.character(all.dists.df$hab2[i])
  sorted.habs <- sort(c(hab1,hab2))
  all.dists.df$hab1[i] = sorted.habs[1]
  all.dists.df$hab2[i] = sorted.habs[2]
  all.dists.df$habs[i] = paste(sorted.habs[1], sorted.habs[2], sep = "-")
}

all.dists.df %>% 
  ggplot(aes(x = as.numeric(geo), y = as.numeric(com))) +
  # geom_point(data = select(all.dists.df, -hab1, -hab2), 
  #            aes(grouping = NULL), alpha = 0.1, color = "grey") +
  geom_point(alpha = 0.25) +
  geom_smooth(method = 'lm') + 
  facet_grid(hab1~hab2) +
  theme(legend.position = "none")
```

